metrics/timers.go

```diff
@@ -0,0 +1,89 @@
+package metrics
+
+import (
+	"context"
+	"sync"
+	"time"
+
+	"gx/ipfs/QmNVpHFt7QmabuVQyguf8AbkLDZoFh7ifBYztqijYT1Sd2/go.opencensus.io/stats"
+	"gx/ipfs/QmNVpHFt7QmabuVQyguf8AbkLDZoFh7ifBYztqijYT1Sd2/go.opencensus.io/stats/view"
+)
+
+// Registry holds a list of measurements
+type Registry struct {
+	msurMu   sync.Mutex
+	measures map[string]*Float64Timer
+}
+
+// NewRegistry creates a new registry
+func NewRegistry() *Registry {
+	return &Registry{
+		measures: make(map[string]*Float64Timer),
+	}
+}
+
+// NewTimer creates a Float64Timer that wraps an opencensus flat64 measurement.
+// The time defaults to milliseconds.
+func (r *Registry) NewTimer(name, desc string) *Float64Timer {
+	r.msurMu.Lock()
+	defer r.msurMu.Unlock()
+	timer, ok := r.measures[name]
+	if ok {
+		return timer
+	}
+	log.Infof("registering timer: %s", name)
+	fMeasure := stats.Float64(name, desc, stats.UnitMilliseconds)
+	fView := &view.View{
+		Name:        name,
+		Measure:     fMeasure,
+		Description: desc,
+		// [>=0ms, >=25ms, >=50ms, >=75ms, >=100ms, >=200ms, >=400ms, >=600ms, >=800ms, >=1s, >=2s, >=4s, >=8s]
+		Aggregation: view.Distribution(25, 50, 75, 100, 200, 400, 600, 800, 1000, 2000, 4000, 8000),
+	}
+	if err := view.Register(fView); err != nil {
+		// a panic here indicates a developer error when creating a view.
+		// Since this method is called in init() methods, this panic when hit
+		// will cause running the program to fail immediately.
+		panic(err)
+	}
+
+	timer = &Float64Timer{
+		measure: fMeasure,
+		view:    fView,
+	}
+	r.measures[name] = timer
+	return timer
+}
+
+// Float64Timer contains a opencensus measurement and view
+type Float64Timer struct {
+	measure *stats.Float64Measure
+	view    *view.View
+}
+
+// Begin starts a timer and returns a Stopwatch.
+func (t *Float64Timer) Begin(ctx context.Context) *Stopwatch {
```

Comment:
Thinking of a better name, open to suggestions -- maybe `Time`?

Reply:
`Start` would be symmetric with `Stop` and natural for a stopwatch.